app/src/main/java/io/apicurio/registry/rest/ArtifactsResourceImpl.java

```diff
@@ -225,23 +258,21 @@ public void testUpdateArtifact(String artifactId, ArtifactType xRegistryArtifact
     }
 
     /**
-     * @see io.apicurio.registry.rest.ArtifactsResource#createArtifact(io.apicurio.registry.types.ArtifactType, java.lang.String, java.io.InputStream)
+     * @see io.apicurio.registry.rest.ArtifactsResource#createArtifact(io.apicurio.registry.types.ArtifactType, java.lang.String, io.apicurio.registry.rest.beans.IfExistsType, java.io.InputStream)
      */
     @Override
-    public CompletionStage<ArtifactMetaData> createArtifact(ArtifactType xRegistryArtifactType, String xRegistryArtifactId,
-                                                            InputStream data) {
-        String artifactId = xRegistryArtifactId;
-        if (artifactId == null || artifactId.trim().isEmpty()) {
-            artifactId = idGenerator.generate();
+    public CompletionStage<ArtifactMetaData> createArtifact(ArtifactType xRegistryArtifactType,
+            String xRegistryArtifactId, IfExistsType ifExists, InputStream data) {
+
+        final CompletionStage<ArtifactMetaData> alreadyExistingArtifactResult = handleIfExists(
+                xRegistryArtifactType, xRegistryArtifactId, ifExists, data);
+
+        if ( null != alreadyExistingArtifactResult) {
+
+            return alreadyExistingArtifactResult;
         }
-        ContentHandle content = ContentHandle.create(data);
 
-        ArtifactType artifactType = determineArtifactType(content, xRegistryArtifactType, request);
-        rulesService.applyRules(artifactId, artifactType, content, RuleApplicationType.CREATE);
-        String finalArtifactId = artifactId;
-        return storage.createArtifact(artifactId, artifactType, content)
-                      .thenCompose(amdd -> indexArtifact(finalArtifactId, content, amdd))
-                      .thenApply(dto -> DtoUtil.dtoToMetaData(finalArtifactId, artifactType, dto));
+        return handleNotExistingArtifactCreation(xRegistryArtifactType, xRegistryArtifactId, data);
```

Comment:
I think this introduces a race condition, but I'm not sure if we should care.  What do you think?  If two attempts to create the same artifact happen at the same time, they might **both** return `null` from `handleIfExists` and then both try to do the create.  The result would be a failure for one of them even if they passed something like `RETURN` for the new query param.

An alternative could be to actually go ahead and try the create.  Then if it fails handle the "if exists" logic at that point.  Thoughts?

Reply:
Yes, I agree that the race condition exists. Another option would be to sync the `handleIfExists` method, but I don't really like that solution.

I think I prefer the option of catching the `ArtifactAlreadyExistsException` at a higher level (instead of storage level) and at that point apply the if exists logic.

Reply:
> Another option would be to sync 
What about in cluster?

Reply:
Not an option then. That's one of the reasons for me to prefer the exception handling.

Reply:
Do we need a storage-specific feature for this?  In-memory and JPA will work if we take the approach of handling the `ArtifactAlreadyExistsException`.  But I suspect Streams might not work if multiple requests to create an artifact come in at the same time to different registry pods.  Am I correct on that, @alesj ?

If so, is there a solution that will work for Streams if we push the functionality **into** the storage?

Reply:
Having a look at the streams storage code, this situation is being handled and also throwing an `ArtifactAlreadyExistsException`.

Reply:
OK if that's the case - then I'd vote for the approach where we try to create it in the JAX-RS layer and then handle the `ArtifactAlreadyExistsException` ... reacting based on the new query param.  Sound good @carlesarnal ?

Reply:
Sure, `ArtifactAlreadyExistException` handling approach.

Reply:
To add to this wrt Streams (and Kafka) ... the problem there is that it's all async -- when the msg/action hits storage -- as we handle this via Kafka producer/consumer.
We do have some initial checks, and we ignore dups when consuming the message, but afais, that's the best we can do; e.g. no point in throwing and exception when things are async.
For Streams I do try a bit better, with that remote Function call, but that's still async -- as you'll get callback only when things are consumed.

Reply:
Sorry, I was thinking about earlier stages of the creation operation like this [one](https://github.com/Apicurio/apicurio-registry/blob/master/storage/streams/src/main/java/io/apicurio/registry/streams/StreamsRegistryStorage.java#L289). If the operation fails/clashes at the consumer level, right, no point in throwing exceptions.